From 134ea87c489325768c0a06c3b029e76a7ffdb8fb Mon Sep 17 00:00:00 2001
From: shxrrydw <sherrydrv@gmail.com>
Date: Tue, 29 Sep 2026 15:34:38 +0200
Subject: [PATCH] server: Don't clip other process children to layered windows.

To prevent other processes from drawing into layered windows updated
with UpdateLayeredWindow, win32u reports a 1x1 surface for them, and
all their children were being clipped to it, so nothing drawn into
them ended up showing.

Based on patches from Lily.

App: Wuthering Waves (3513350).
---
 server/window.c | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/server/window.c b/server/window.c
index f7ae76acf48..2c6291f9988 100644
--- a/server/window.c
+++ b/server/window.c
@@ -1315,12 +1315,13 @@ static struct region *clip_children( struct window *parent, struct window *last,
 
 
 /* set the region to the client rect clipped by the window rect, in parent-relative coordinates */
-static void set_region_client_rect( struct region *region, struct window *win )
+static void set_region_client_rect( struct region *region, struct window *win, struct process *process )
 {
     struct rectangle rect;
 
     intersect_rect( &rect, &win->window_rect, &win->client_rect );
-    intersect_rect( &rect, &rect, &win->surface_rect );
+    if (!(win->paint_flags & PAINT_HAS_LAYERED_SURFACE) || !win->thread || win->thread->process == process)
+        intersect_rect( &rect, &rect, &win->surface_rect );
     set_region_rect( region, &rect );
 }
 
@@ -1438,6 +1439,7 @@ static void hwnd_dmabuf_fill_frame_info( struct window *host, struct window *win
 /* compute the visible region of a window, in window coordinates */
 static struct region *get_visible_region( struct window *win, unsigned int flags )
 {
+    struct process *process = win->thread ? win->thread->process : NULL;
     struct region *tmp = NULL, *region;
     int offset_x, offset_y;
 
@@ -1468,7 +1470,7 @@ static struct region *get_visible_region( struct window *win, unsigned int flags
 
     if ((flags & DCX_PARENTCLIP) && !is_desktop_window( win->parent ))
     {
-        set_region_client_rect( region, win->parent );
+        set_region_client_rect( region, win->parent, process );
         offset_region( region, -win->parent->client_rect.left, -win->parent->client_rect.top );
     }
     else if (flags & DCX_WINDOW)
@@ -1478,7 +1480,7 @@ static struct region *get_visible_region( struct window *win, unsigned int flags
     }
     else
     {
-        set_region_client_rect( region, win );
+        set_region_client_rect( region, win, process );
         if (win->win_region && !intersect_window_region( region, win )) goto error;
     }
 
@@ -1509,7 +1511,7 @@ static struct region *get_visible_region( struct window *win, unsigned int flags
             offset_x += win->client_rect.left;
             offset_y += win->client_rect.top;
             offset_region( region, win->client_rect.left, win->client_rect.top );
-            set_region_client_rect( tmp, win );
+            set_region_client_rect( tmp, win, process );
             if (win->win_region && !intersect_window_region( tmp, win )) goto error;
             if (!intersect_region( region, region, tmp )) goto error;
             if (is_region_empty( region )) break;
